### PR TITLE
lint: Add linter for circular dependencies

### PR DESCRIPTION
**Pull request by Empact**
Protects against added circular depencies, makes it explicit in the
code when circular dependencies have been removed.

Modeled after EXPECTED_BOOST_INCLUDES in lint-includes.sh

Example output:
```
$ test/lint/lint-circular-dependencies.sh 
A new circular dependency in the form of "qt/paymentserver -> qt/walletmodel -> qt/paymentserver" appears to have been introduced.

$ echo $?
1
$ test/lint/lint-circular-dependencies.sh 
Good job! The circular dependency "Fake" is no longer present.
Please remove it from EXPECTED_CIRCULAR_DEPENDENCIES in test/lint/lint-circular-dependencies.sh
to make sure this circular dependency is not accidentally reintroduced.

$ echo $?
1
$ test/lint/lint-circular-dependencies.sh 
$ echo $?
0

```

### CODE REVIEW

**Comment by practicalswift** [comment]
Very very nice! This will help us get rid of circular dependencies permanently! Thanks for doing this.

ACK 54f16f23e886f6c37189aa53c86cdb0cefa809ea

File: test/lint/lint-circular-dependencies.sh [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    "qt/transactiontablemodel -> qt/walletmodel -> qt/transactiontablemodel"
+    "qt/walletmodel -> qt/walletmodeltransaction -> qt/walletmodel"
+    "rpc/rawtransaction -> wallet/rpcwallet -> rpc/rawtransaction"
+    "txmempool -> validation -> txmempool"
+    "validation -> validationinterface -> validation"
+    "wallet/coincontrol -> wallet/wallet -> wallet/coincontrol"
+    "wallet/fees -> wallet/wallet -> wallet/fees"
+    "wallet/rpcwallet -> wallet/wallet -> wallet/rpcwallet"
+    "wallet/wallet -> wallet/walletdb -> wallet/wallet"
+    "policy/fees -> policy/policy -> validation -> policy/fees"
```
[the remark this answers is not in the document]
Comment (Empact) [reply test/lint/lint-circular-dependencies.sh]: These are sorted consistent with the output of `contrib/devtools/circular-dependencies.py`, which is deterministic. https://github.com/bitcoin/bitcoin/pull/13228#issuecomment-389701793

File: test/lint/lint-circular-dependencies.sh [hunk 2]
```diff
@@ -0,0 +1,85 @@
+#!/usr/bin/env bash
+#
+# Copyright (c) 2018 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+#
+# Make sure only lowercase alphanumerics (a-z0-9), underscores (_),
```
[the remark this answers is not in the document]
Comment (Empact) [reply test/lint/lint-circular-dependencies.sh]: Updated description. Thanks!

**Comment by practicalswift** [comment]
ACK 5c613aadd64453c75cb2373c6fcc1326c3cf0b7a

**Comment by laanwj** [comment]
utACK 5c613aadd64453c75cb2373c6fcc1326c3cf0b7a

**Pull request merged**